Mathlib/Combinatorics/SimpleGraph/Copy.lean

```diff
@@ -348,14 +355,204 @@ lemma isIndContained_iff_exists_iso_subgraph :
+
+@[simp] lemma copyCount_eq_zero : G.copyCount H = 0 ↔ ¬ G ⊑ H := by
+  simp [copyCount, -nonempty_subtype, isContained_iff_exists_iso_subgraph, card_pos,
+    filter_eq_empty_iff]
+
+@[simp] lemma copyCount_pos : 0 < G.copyCount H ↔ G ⊑ H := by
+  simp [copyCount, -nonempty_subtype, isContained_iff_exists_iso_subgraph, card_pos,
+    filter_nonempty_iff]
+
+end CopyCount
+
+section LabelledCopyCount
+variable [Fintype V] [Fintype W]
+
+/-- `G.labelledCopyCount H` is the number of labelled copies of `G` in `H`. See
+`SimpleGraph.copyCount` for the number of unlabelled copies. -/
+noncomputable def labelledCopyCount (G : SimpleGraph V) (H : SimpleGraph W) : ℕ := by
+  classical exact Fintype.card (Copy G H)
+
+@[simp] lemma labelledCopyCount_of_isEmpty [IsEmpty V] (G : SimpleGraph V) (H : SimpleGraph W) :
+    G.labelledCopyCount H = 1 := by
+  convert Fintype.card_unique
+  exact { default := ⟨default, isEmptyElim⟩, uniq := fun _ ↦ Subsingleton.elim _ _ }
+
+@[simp] lemma labelledCopyCount_eq_zero : G.labelledCopyCount H = 0 ↔ ¬ G ⊑ H := by
+  simp [labelledCopyCount, IsContained, Fintype.card_eq_zero_iff, isEmpty_subtype]
+
+@[simp] lemma labelledCopyCount_pos : 0 < G.labelledCopyCount H ↔ G ⊑ H := by
+  simp [labelledCopyCount, IsContained, Fintype.card_pos_iff]
+
+/-- There's more labelled copies of `H` of-`G` than unlabelled ones. -/
+lemma copyCount_le_labelledCopyCount : G.copyCount H ≤ G.labelledCopyCount H := by
+  classical
+  rw [copyCount, ← Fintype.card_coe]
+  refine Fintype.card_le_of_surjective (fun f ↦ ⟨f.toSubgraph, ?_⟩) ?_
+  · simpa using ⟨f.isoToSubgraph⟩
+  · simpa [Surjective, -Copy.range_toSubgraph] using fun H' e ↦ Copy.toSubgraph_surjOn ⟨e⟩
+
+end LabelledCopyCount
+
+/-!
+#### Induced copies
+
+TODO
 
-@[simp] lemma compl_isIndContained_compl : Gᶜ ⊴ Hᶜ ↔ G ⊴ H :=
-  Embedding.complEquiv.symm.nonempty_congr
+### Killing a subgraph
 
-protected alias ⟨IsIndContained.of_compl, IsIndContained.compl⟩ := compl_isIndContained_compl
+An important aspect of graph containment is that we can remove not too many edges from a graph `H`
+to get a graph `H'` that doesn't contain `G`.
+
+#### Killing not necessarily induced copies
+
+`SimpleGraph.kill G H` is a subgraph of `H` where an edge was removed from each copy of `G` in `H`.
+By construction, it doesn't contain `G` and has at most the number of copies of `G` edges less than
+`H`
+-/
+
+private lemma aux (hG : G ≠ ⊥) {H' : H.Subgraph} :
+    Nonempty (G ≃g H'.coe) → H'.edgeSet.Nonempty := by
+  obtain ⟨e, he⟩ := edgeSet_nonempty.2 hG
+  rw [← Subgraph.image_coe_edgeSet_coe]
+  exact fun ⟨f⟩ ↦ Set.Nonempty.image _ ⟨_, f.map_mem_edgeSet_iff.2 he⟩
+
+/-- `G.kill H` is a subgraph of `H` where an edge from every subgraph isomorphic to `G` was removed.
+As such, it is a big subgraph of `H` that does not contain any subgraph isomorphic to `G`, unless
+`G` had no edges to start with. -/
+noncomputable irreducible_def kill (G : SimpleGraph α) (H : SimpleGraph W) : SimpleGraph W := by
```

Comment:
Let's go with `killCopies`, since there's many things you could kill in a graph!

Reply:
Ah! I wasn't quite sure what to call that, but `killCopies` is a great name